### ISSUE TITLE
Custom Layer Issue

### ISSUE THREAD

**Issue by julie-jiang** [fchollet/keras#7401]
I'm trying to create a lambda layer that will perform some deterministic masking (I'm not talking about the Keras Masking layer) before pumping out the final output. This is what I have so far:

```
def binary_mask(x):
    # Mask is half the size of x. 
    # 1 if pred1 > pred2 element-wise, 0 otherwise.

    pred1 = x(:, :units)
    pred2 = x(: units:)
    mask = K.greater(pred1, pred2)
    mask = K.cast(mask, ,K.floatx())
    return mask

def mask_output_shape(input_shape):
    return (input_shape[0], units)
```
And this is how I create the layer with the functional API:
```
outputs = Lambda(binary_mask, output_shape=mask_output_shape)(inputs) 
```
I'm getting a very weird error coming from my optimizer. It appears that my gradient is `None`.

However, if I replace my current code in `binary_mask` with any of the example Lambda layer implementations, it works fine, e.g. `return x`. What am I doing wrong here?

**Comment by mrTsjolder** [comment]
You have created a layer with outputs `pred1 > pred2`. This is not differentiable, hence the error.

Assuming that you actually want something like `x if pred1 > pred2 else 0` in the end, you could simply use `return mask * x`. This should be differentiable (enough) to have keras optimise it for you.

If you really need the 0/1 outputs, it might help to use `K.switch(mask, 1.0, 0.0)`, but this might not work with the Tensorflow backend (`tf.cond` only accepts scalar conditions). With `tf.where`, you could have the correct behaviour, but I do not find a way to access this functionality through the Keras backend interface. I am not sure, but it might be that an explicit split is necessary for your function to be differentiable.

Could anybody verify that there is nothing like `K.where` for element-wise conditions in the Keras backend. It seems like this is a useful function missing there.

**Comment by julie-jiang** [comment]
I'm afraid I don't completely understand why it would be not differentiable. My target vector is full of 0s and 1s. So I want my input vector to match the distribution of 0s and 1s as close as possible. Could you please explain a bit more?

**Comment by mrTsjolder** [comment]
The 0s and 1s are in fact booleans that have been casted to floats, i.e. it is just 

    f(pred1, pred2): return pred1 > pred2

It is hard to formulate a sensible derivative for this function.
If you use `K.switch` (or `tf.where`), you compute 

    f(pred1, pred2): return 1 if pred1 > `pred2 else 0

This is differentiable in the sense that each part of the piecewise derivative exists (both are constant functions, so both have derivative 0). Obviously this does not make much sense and you probably want to multiply your mask with something else, which will lead to more useful derivatives.

However, I am not sure whether this will fix your problem, because it might be possible that the backend you are using is smart enough to discover that the cast from boolean to float is actually some kind of switch. It would be nice if you could confirm my theory...
